src/iocp/port.rs

```diff
@@ -0,0 +1,317 @@
+
+        if result == 0 {
+            return Err(io::Error::last_os_error());
+        }
+
+        if skip_set_event_on_handle {
+            // Set the skip event on handle.
+            let result = unsafe {
+                SetFileCompletionNotificationModes(handle as _, FILE_SKIP_SET_EVENT_ON_HANDLE as _)
+            };
+
+            if result == 0 {
+                return Err(io::Error::last_os_error());
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Post a completion packet to this port.
+    pub(super) fn post(&self, bytes_transferred: usize, id: usize, packet: T) -> io::Result<()> {
+        let result = unsafe {
+            PostQueuedCompletionStatus(
+                self.handle,
+                bytes_transferred
+                    .try_into()
+                    .expect("too many bytes transferred"),
+                id,
+                T::into_ptr(packet),
+            )
+        };
+
+        if result == 0 {
+            Err(io::Error::last_os_error())
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Wait for completion packets to arrive.
+    pub(super) fn wait(
+        &self,
+        packets: &mut Vec<OverlappedEntry<T>>,
+        timeout: Option<Duration>,
+    ) -> io::Result<usize> {
+        // Drop the current packets.
+        packets.clear();
+
+        let mut count = MaybeUninit::<u32>::uninit();
+        let timeout = timeout.map_or(INFINITE, dur2timeout);
+
+        let result = unsafe {
+            GetQueuedCompletionStatusEx(
+                self.handle,
+                packets.as_mut_ptr() as _,
+                packets.capacity().try_into().expect("too many packets"),
+                count.as_mut_ptr(),
+                timeout,
+                0,
+            )
+        };
+
+        if result == 0 {
+            let io_error = io::Error::last_os_error();
+            if io_error.kind() == io::ErrorKind::TimedOut {
+                Ok(0)
+            } else {
+                Err(io_error)
+            }
+        } else {
+            let count = unsafe { count.assume_init() };
+            unsafe {
+                packets.set_len(count as _);
+            }
+            Ok(count as _)
+        }
+    }
+}
+
+/// An `OVERLAPPED_ENTRY` resulting from an I/O completion port.
+#[repr(transparent)]
+pub(super) struct OverlappedEntry<T: CompletionHandle> {
+    /// The underlying entry.
+    entry: OVERLAPPED_ENTRY,
+
+    /// We own the status block.
+    _marker: PhantomData<T>,
+}
+
+impl<T: CompletionHandle> fmt::Debug for OverlappedEntry<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("OverlappedEntry { .. }")
+    }
+}
+
+impl<T: CompletionHandle> OverlappedEntry<T> {
+    /// Convert into the completion packet.
+    pub(super) fn into_packet(self) -> T {
+        let packet = unsafe { self.packet() };
+        std::mem::forget(self);
+        packet
+    }
+
+    unsafe fn packet(&self) -> T {
```

Comment:
what are the safety requirements of this function?

Reply:
It should only be called once, since it moved out the `T`, like `ManuallyDrop::take()`. I've clarified this in the docs.